### PR TITLE
Added functional API to builders. Resolved #26

### LINKED ISSUE #26: Add a suport for building Decision model using lambda expressions.

### LINKED ISSUE #27: Rewrite the YAML example to Java lambda-based API as an additional one in README.md file.

### PR DESCRIPTION
**Pull request by stlhrt**
Resolved #26
Resolved #27

### CODE REVIEW

File: src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java [hunk 1]
```diff
@@ -72,70 +73,97 @@ public ExpressionType getExpressionType() {
...
         protected void initProduct() {
             this.product = new Decision();
         }
 
-        public Builder<P> id(String id) {
+        public B id(String id) {
             this.product.id = id;
 
-            return this;
+            return (B) this;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java]: here will be unchecked cast

File: src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java [hunk 2]
```diff
@@ -72,70 +73,97 @@ public ExpressionType getExpressionType() {
...
 
-    public static final class Builder<P extends AbstractBuilder> extends AbstractBuilder<Decision> {
-
-        private Builder() {
-        }
+    public static Builder builder() {
+        return new Builder();
+    }
 
+    public static abstract class DecisionBuilder<B extends DecisionBuilder<B>> extends AbstractBuilder<Decision> {
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java]: this class should be private be private

File: src/main/java/org/powerflows/dmn/engine/model/decision/rule/Rule.java [hunk 3]
```diff
@@ -50,59 +51,85 @@ public String getDescription() {
...
-            return InputEntry.builder(this, inputEntryConsumer);
+            return this;
         }
+    }
 
-        public OutputEntry.Builder<Builder<P>> withOutputEntries() {
-            final Consumer<OutputEntry> outputEntryConsumer = outputEntry -> this.product.outputEntries.add(outputEntry);
+    public static final class FluentBuilder<P extends AbstractBuilder> extends RuleBuilder<FluentBuilder<P>> {
+        private P parentBuilder;
+        protected final Consumer<Rule> callback;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/model/decision/rule/Rule.java]: should be private

File: src/main/java/org/powerflows/dmn/engine/model/decision/rule/Rule.java [hunk 4]
```diff
@@ -50,59 +51,85 @@ public String getDescription() {
...
 
-            return OutputEntry.builder(this, outputEntryConsumer);
+        public FluentBuilder(final P parentBuilder, final Consumer<Rule> callback) {
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/model/decision/rule/Rule.java]: constructor should be private

File: src/main/java/org/powerflows/dmn/engine/model/decision/rule/Rule.java [hunk 5]
```diff
@@ -50,59 +51,85 @@ public String getDescription() {
...
         @Override
         protected void initProduct() {
             this.product = new Rule();
         }
 
-        public Builder<P> description(String description) {
+        public B description(String description) {
             this.product.description = description;
 
+            return (B) this;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/model/decision/rule/Rule.java]: I think unchecked cast. Please apply to other similar classes.

File: src/test/groovy/org/powerflows/dmn/engine/model/decision/DecisionSpec.groovy [hunk 6]
```diff
@@ -141,6 +143,243 @@ class DecisionSpec extends Specification {
...
+        rule2InputEntry2.getName() == someRule2InputEntry2Name
+
+        final OutputEntry rule2OutputEntry1 = rule2.getOutputEntries().get(0)
+        rule2OutputEntry1.getName() == someRule2OutputEntry1Name
+
+        final OutputEntry rule2OutputEntry2 = rule2.getOutputEntries().get(1)
+        rule2OutputEntry2.getName() == someRule2OutputEntry2Name
+    }
+
+    void 'should build table with fluent API using lambdas'() {
```
Comment (mariuszkumor) [inline src/test/groovy/org/powerflows/dmn/engine/model/decision/DecisionSpec.groovy]: should build decision

Comment (mariuszkumor) [reply src/test/groovy/org/powerflows/dmn/engine/model/decision/DecisionSpec.groovy]: please change also above

File: README.md [hunk 7]
```diff
@@ -195,7 +195,97 @@ Decision decision = Decision.builder()
                             .and()
                         .end()
                     .end()
+                .build()
+```
+
+## Java with functional API
+
+```java
+final Decision decision = Decision.builder()
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply README.md]: ok right

File: src/main/java/org/powerflows/dmn/engine/model/builder/AbstractBuilder.java [hunk 8]
```diff
@@ -16,22 +16,27 @@
 
 package org.powerflows.dmn.engine.model.builder;
 
-public abstract class AbstractBuilder<T> {
-
+public abstract class AbstractBuilder<T> implements Builder<T> {
     protected T product;
 
-    {
+    public AbstractBuilder() {
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/engine/model/builder/AbstractBuilder.java]: this is way better than static block

File: src/main/java/org/powerflows/dmn/engine/model/builder/AbstractBuilder.java [hunk 9]
```diff
@@ -16,22 +16,27 @@
...
 
     protected T assembleProduct() {
         return this.product;
     }
 
+    @Override
+    public final T build() {
+        final T temp = assembleProduct();
+        if (temp == null) {
+            throw new IllegalStateException("Only single build() call is allowed.");
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/engine/model/builder/AbstractBuilder.java]: when someone calls build more than one which is invalid API usage

File: src/main/java/org/powerflows/dmn/engine/model/builder/Builder.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package org.powerflows.dmn.engine.model.builder;
+
+public interface Builder<T> {
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/engine/model/builder/Builder.java]: ok

File: src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java (hunk 2 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java]: hmmm... possibly...

File: src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java (hunk 1 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/engine/model/decision/Decision.java]: true, but there is no other way, the only implementation is here so the cast is always valid

**Pull request closed without merging**